modules/fedimint-ln-client/src/pay.rs

```diff
@@ -251,50 +255,19 @@ impl LightningPayFunded {
         Ok(preimage[1..length - 1].to_string())
     }
 
-    async fn transition_pay_timeout(
-        old_state: LightningPayStateMachine,
-        global_context: DynGlobalClientContext,
-        contract_id: ContractId,
-    ) -> LightningPayStateMachine {
-        // TODO: Retry contacting gateway
-
-        let contract = global_context
-            .api()
-            .get_outgoing_contract(contract_id)
-            .await;
-        let timelock = match contract {
-            Ok(contract) => contract.contract.timelock,
-            Err(_) => {
-                return LightningPayStateMachine {
-                    common: old_state.common,
-                    state: LightningPayStates::Failure(
-                        "Failed to retrieve OutgoingContract".to_string(),
-                    ),
-                }
-            }
-        };
-
-        LightningPayStateMachine {
-            common: old_state.common,
-            state: LightningPayStates::Refundable(LightningPayRefundable {
-                contract_id,
-                block_timelock: timelock,
-            }),
-        }
-    }
-
     async fn transition_outgoing_contract_execution(
         old_state: LightningPayStateMachine,
         global_context: DynGlobalClientContext,
-        result: Result<String, GatewayPayError>,
         contract_id: ContractId,
+        gateway: LightningGateway,
+        payload: PayInvoicePayload,
     ) -> LightningPayStateMachine {
-        match result {
+        match Self::gateway_pay_invoice(gateway, payload).await {
```

Comment:
State transitions should ideally not make API calls since they can't run in parallel and e.g. the API being slow would block all other state machines from making progress. What is the reason for moving the fn here? It will still be run multiple times if the client is restarted at the wrong time, so the API endpoint has to be idempotent anyway (tbf, doing the call here will make this less frequent).

Reply:
As discussed offline, this is difficult to fix without client NG on the gateway side. Made an issue: https://github.com/fedimint/fedimint/issues/2496